cmd/containerboot/main_test.go

```diff
@@ -0,0 +1,438 @@
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"golang.org/x/sys/unix"
+	"tailscale.com/ipn/ipnstate"
+)
+
+func TestContainerBoot(t *testing.T) {
+	d, err := os.MkdirTemp("", "containerboot")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	lapi := localAPI{FSRoot: d}
+	if err := lapi.Start(); err != nil {
+		t.Fatal(err)
+	}
+	defer lapi.Close()
+
+	kube := kubeServer{FSRoot: d}
+	if err := kube.Start(); err != nil {
+		t.Fatal(err)
+	}
+	defer kube.Close()
+
+	for _, path := range []string{"var/lib", "usr/bin", "tmp"} {
+		if err := os.MkdirAll(filepath.Join(d, path), 0700); err != nil {
+			t.Fatal(err)
+		}
+	}
+	if err := os.WriteFile(filepath.Join(d, "usr/bin/tailscaled"), fakeTailscaled, 0700); err != nil {
+		t.Fatal(err)
+	}
+	if err := os.WriteFile(filepath.Join(d, "usr/bin/tailscale"), fakeTailscale, 0700); err != nil {
+		t.Fatal(err)
+	}
+
+	boot := filepath.Join(d, "containerboot")
+	if err := exec.Command("go", "build", "-o", boot, "tailscale.com/cmd/containerboot").Run(); err != nil {
+		t.Fatalf("Building containerboot: %v", err)
+	}
+
+	argFile := filepath.Join(d, "args")
+
+	lapi.Reset()
+	kube.Reset()
+
+	cmd := exec.Command(boot)
+	cmd.Env = []string{
+		fmt.Sprintf("PATH=%s/usr/bin:%s", d, os.Getenv("PATH")),
+		fmt.Sprintf("TS_TEST_RECORD_ARGS=%s", argFile),
+		fmt.Sprintf("TS_TEST_SOCKET=%s", lapi.Path),
+		fmt.Sprintf("TS_SOCKET=%s", filepath.Join(d, "tmp/tailscaled.sock")),
+	}
+	cbOut := &lockingBuffer{}
+	cmd.Stderr = cbOut
+	if err := cmd.Start(); err != nil {
+		t.Fatalf("starting containerboot: %v", err)
+	}
+	defer func() {
+		cmd.Process.Signal(unix.SIGTERM)
+		cmd.Process.Wait()
+	}()
+
+	want := `
+/usr/bin/tailscaled --socket=/tmp/tailscaled.sock --state=mem: --statedir=/tmp --tun=userspace-networking
+/usr/bin/tailscale --socket=/tmp/tailscaled.sock up --accept-dns=false
+`
+	waitArgs(t, 2*time.Second, d, argFile, want)
+
+	lapi.SetStatus(ipnstate.Status{
+		BackendState: "Running",
+		TailscaleIPs: []netip.Addr{
+			netip.MustParseAddr("100.64.0.1"),
+		},
+	})
+
+	waitLogLine(t, 2*time.Second, cbOut, "Startup complete, waiting for shutdown signal")
+}
+
+type lockingBuffer struct {
+	sync.Mutex
+	b bytes.Buffer
+}
+
+func (b *lockingBuffer) Write(bs []byte) (int, error) {
+	b.Lock()
+	defer b.Unlock()
+	return b.b.Write(bs)
+}
+
+func (b *lockingBuffer) String() string {
+	b.Lock()
+	defer b.Unlock()
+	return b.b.String()
+}
+
+func waitLogLine(t *testing.T, timeout time.Duration, b *lockingBuffer, want string) {
+	deadline := time.Now().Add(timeout)
+	for time.Now().Before(deadline) {
+		for _, line := range strings.Split(b.String(), "\n") {
+			if !strings.HasPrefix(line, "boot: ") {
+				continue
+			}
+			if strings.HasSuffix(line, " "+want) {
+				return
+			}
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+	t.Fatalf("timed out waiting for wanted output line %q. Output:\n%s", want, b.String())
+}
+
+func waitArgs(t *testing.T, timeout time.Duration, root, path, wantArgs string) {
```

Comment:
all args except root are self-explanatory.

Reply:
add a comment?

Reply:
Documented both wait funcs, and renamed `root` to `removeStr`.